lib/phoenix/endpoint.ex

```diff
@@ -909,13 +852,38 @@ defmodule Phoenix.Endpoint do
 
   """
   defmacro socket(path, module, opts \\ []) do
-    module = Macro.expand(module, %{__CALLER__ | function: {:__handler__, 2}})
+    module = Macro.expand(module, __CALLER__)
+    websocket = Keyword.get(opts, :websocket, true)
+    longpoll = Keyword.get(opts, :longpoll, false)
 
     quote do
+      unquote(socket_route(path, module, websocket, Phoenix.Transports.WebSocket))
+      unquote(socket_route(path, module, longpoll, Phoenix.Transports.LongPoll))
```

Comment:
WebSocket and long polling are now exposed directly as plugs within Endpoint, matched via `Plug.Router.match/5` that we return directly from this macro